core/mixins/LogSupport.js

```diff
@@ -0,0 +1,98 @@
+    return applyMixin(C, {
+        name: 'LogSupport',
+
+        provides: {
+
+            /**
+             * Surround a function's execution with log messages, and timing info.
+             *
+             * @param {(String[]|String)} msgs -- One or more strings to form a compound log message.
+             * @param {function} fn - function to be tracked by this log message.
+             * @returns {*} - value returned by fn
+             */
+            withLog(msgs, fn) {
+                return this.xhLoggedDo({level: 'log', full: true, msgs, fn});
+            },
+
+            /**
+             * Track a function's execution with a single log message, and timing info.
+             *
+             * @param {(String[]|String)} msgs -- One or more strings to form a compound log message.
+             * @param {function} fn - function to be tracked by this log message.
+             * @returns {*} - value returned by fn
+             */
+            withShortLog(msgs, fn) {
+                return this.xhLoggedDo({level: 'log', full: false, msgs, fn});
+            },
+
+            /**
+             * Surround a function's execution with debug messages, and timing info.
+             *
+             * @param {(String[]|String)} msgs -- One or more strings to form a compound log message.
+             * @param {function} fn - function to be tracked by this log message.
+             * @returns {*} - value returned by fn
+             */
+            withDebug(msgs, fn) {
+                return this.xhLoggedDo({level: 'debug', full: true, msgs, fn});
+            },
+
+            /**
+             * Track a function's execution with a single debug message, and timing info.
+             *
+             * @param {(String[]|String)} msgs -- One or more strings to form a compound log message.
+             * @param {function} fn - function to be tracked by this log message.
+             * @returns {*} - value returned by fn
+             */
+            withShortDebug(msgs, fn) {
+                return this.xhLoggedDo({level: 'debug', full: false, msgs, fn});
+            },
+            
+            //----------------------------------
+            // Implementation
+            //----------------------------------
+            xhLoggedDo({level, full, msgs, fn}) {
+
+                msgs = castArray(msgs);
+
+                const start = Date.now(),
+                    msg = msgs.join(' | ');
+
+                if (full) this.xhLogAtLevel(level, `${msg} | started`);
+
+                let ret, elapsed;
+                try {
+                    ret = fn();
+                } catch (e) {
+                    elapsed = Date.now() - start;
+                    this.xhLogAtLevel(level, `${msg} | failed - ${e.message || e.name || 'Unknown error'} | ${elapsed}`);
+                    throw e;
+                }
+
+                elapsed = Date.now() - start;
+                this.xhLogAtLevel(level, `${msg} | completed | ${elapsed}`);
```

Comment:
Great feedback re: async.  I think we can handle it in a single async version (see last commit) -- not sure how to name it,  could have two versions, just trying to avoid too many permutations...